File: python/src/roar_sdk/hub_auth.py
# -*- coding: utf-8 -*-
"""ROAR Protocol — Hub challenge-response authentication.

Provides ChallengeStore for proof-of-possession registration:
  1. Hub issues a short-lived nonce challenge tied to the registrant's DID.
  2. Client signs the nonce with its Ed25519 private key.
  3. Hub verifies the signature, then admits the agent card.

This store is single-process. Distributed deployments should replace it with
Redis SETNX with TTL for atomic challenge issuance.
"""

from __future__ import annotations

import secrets
import time
from dataclasses import dataclass, field
from typing import Dict, Optional


@dataclass
class PendingChallenge:
    challenge_id: str
    did: str
    nonce: str
    expires_at: float
    public_key: str   # hex-encoded Ed25519 public key from registration request
    card: dict = field(default_factory=dict)  # raw AgentCard dict to register on success


class ChallengeStore:
    """In-memory store for pending registration challenges.

    Nonces expire after NONCE_TTL_SECONDS. Replayed challenge_ids are rejected
    because ``consume`` deletes the entry on first retrieval.

    This store is single-process; distributed deployments should replace with
    Redis SETNX with TTL for atomic challenge issuance.
    """

    NONCE_TTL_SECONDS = 30.0
    MAX_PENDING = 1000

    def __init__(self) -> None:
        self._pending: Dict[str, PendingChallenge] = {}

    def issue(self, did: str, public_key: str, card: dict) -> PendingChallenge:
        """Issue a new challenge for *did*.  Evicts expired entries first."""
        self._evict_expired()
        if len(self._pending) >= self.MAX_PENDING:
            raise RuntimeError("Too many pending challenges — server busy")
        challenge_id = secrets.token_hex(16)
        nonce = secrets.token_hex(32)
        challenge = PendingChallenge(
            challenge_id=challenge_id,
            did=did,
            nonce=nonce,
            expires_at=time.time() + self.NONCE_TTL_SECONDS,
            public_key=public_key,
            card=card,
        )
        self._pending[challenge_id] = challenge
        return challenge

    def consume(self, challenge_id: str) -> Optional[PendingChallenge]:
        """Return and DELETE the challenge (prevents replay).

        Returns None if the challenge has expired or was never issued.
        """
        self._evict_expired()
        return self._pending.pop(challenge_id, None)

    def _evict_expired(self) -> None:
        now = time.time()
        expired = [cid for cid, c in self._pending.items() if c.expires_at < now]
        for cid in expired:
            del self._pending[cid]

File: ts/src/hub_auth.ts
/**
 * ROAR Protocol — Hub challenge-response authentication (TypeScript).
 * Mirrors python/src/roar_sdk/hub_auth.py.
 *
 * This in-memory store is single-process.
 * Distributed deployments should replace with Redis SETNX + TTL.
 */

import * as crypto from "crypto";

export interface PendingChallenge {
  challenge_id: string;
  did: string;
  nonce: string;
  expires_at: number;
  public_key: string; // hex Ed25519 public key
  card: Record<string, unknown>;
}

export class ChallengeStore {
  static readonly NONCE_TTL_SECONDS = 30.0;
  static readonly MAX_PENDING = 1000;

  private _pending = new Map<string, PendingChallenge>();

  issue(did: string, public_key: string, card: Record<string, unknown>): PendingChallenge {
    this._evictExpired();
    if (this._pending.size >= ChallengeStore.MAX_PENDING) {
      throw new Error("Too many pending challenges — server busy");
    }

    const challenge_id = crypto.randomBytes(16).toString("hex");
    const nonce = crypto.randomBytes(32).toString("hex");
    const expires_at = Date.now() / 1000 + ChallengeStore.NONCE_TTL_SECONDS;

    const ch: PendingChallenge = {
      challenge_id,
      did,
      nonce,
      expires_at,
      public_key,
      card,
    };

    this._pending.set(challenge_id, ch);
    return ch;
  }

  /** Return and DELETE the challenge (prevents replay). */
  consume(challenge_id: string): PendingChallenge | null {
    this._evictExpired();
    const ch = this._pending.get(challenge_id) ?? null;
    if (ch) this._pending.delete(challenge_id);
    return ch;
  }

  private _evictExpired(): void {
    const now = Date.now() / 1000;
    for (const [id, ch] of this._pending.entries()) {
      if (ch.expires_at < now) this._pending.delete(id);
    }
  }
}
